hbp_validation_framework/__init__.py

```diff
@@ -1910,6 +1921,152 @@ def get_model_instance(self, instance_path="", instance_id="", model_id="", alia
         else:
             raise Exception("Error in retrieving model instance. Possibly invalid input data.")
 
+    def download_model_instance(self, instance_path="", instance_id="", model_id="", alias="", version="", target_loc="."):
+        """Download files/directory corresponding to an existing model instance.
+
+        Files/directory corresponding to a model instance to be downloaded. The model
+        instance can be specified in the following ways (in order of priority):
+
+        1. load from a local JSON file specified via `instance_path`
+        2. specify `instance_id` corresponding to model instance in model catalog
+        3. specify `model_id` and `version`
+        4. specify `alias` (of the model) and `version`
+
+        Parameters
+        ----------
+        instance_path : string
+            Location of local JSON file with model instance metadata.
+        instance_id : UUID
+            System generated unique identifier associated with model instance.
+        model_id : UUID
+            System generated unique identifier associated with model description.
+        alias : string
+            User-assigned unique identifier associated with model description.
+        version : string
+            User-assigned identifier (unique for each model) associated with model instance.
+        target_loc : string
+            Directory path (relative/absolute) where files should be downloaded and saved. Default is current location.
+            Existing files, if any, at the target location will be overwritten!
+
+        Returns
+        -------
+        string
+            Absolute path of the downloaded file/directory.
+
+        Note
+        ----
+        Existing files, if any, at the target location will be overwritten!
+
+        Examples
+        --------
+        >>> file_path = model_catalog.download_model_instance(instance_id="a035f2b2-fe2e-42fd-82e2-4173a304263b")
+        """
+
+        model_source = self.get_model_instance(instance_path=instance_path, instance_id=instance_id, model_id=model_id, alias=alias, version=version)["source"]
+        if model_source[-1]=="/":
+            model_source = model_source[:-1]    # remove trailing '/'
+        if not os.path.exists(target_loc):
+            os.makedirs(target_loc) # create directory if it doesn't exist
+        fileList = []
+
+        if model_source.startswith("https://collab.humanbrainproject.eu/#/collab/"):
```

Comment:
In the interests of ease-of-testing, maintainability and extensibility, I strongly suggest moving this functionality into the datastores module, i.e.

- extend `CollabDataStore`, if necessary
- add a `SwiftDataStore` class.

Reply:
Will look into this and see how best it to accommodate it there.

Reply:
Done as per above suggestion. Also moved a few methods from __init__.py to CollabDataStore (ideally I should never have placed them in the former). So a good amount of code clean-up carried out.